### ISSUE TITLE
No connection to x22i

### ISSUE THREAD

**Issue by samkenw** [cagnulein/QZCompanionNordictrackTreadmill#65]
Hi there, 

Trying to get this to run with my Nordictrack x22i. 

**What I did:**
Installed companion via browser (no windows pc available).
In the companion app I selected "other".
In the QZ app (iOS) added the IP address of the treadmill.

**Issue:**
In the iOS app I am seeing the data fields - but nothing is changing. 

**Logs:**
In the log dump in the companion app on the treadmill I get: 

2023-11-03 17:59:18.755 Service started
2023-11-03 17:59:18.765 Service onCreate
2023-11-03 17:59:18.765 Service postDelayed
2023-11-03 17:59:18.766 Service started
2023-11-03 17:59:19.267 Service run
2023-11-03 17:59:19.276 /sdcard/.wolflogs/
2023-11-03 17:59:19.276 lastModifiedFile /scard/wolflogs/2023-11-03 _logs.txt
2023-11-03 17:59:19.277 string: /sdeard/wolflogs/2023-11-03_logs.txt
2023-11-03 17:59:19.277 Parsing /sdcard/.wolflogs/2023-11-03_logs.txt
2023-11-03 17:59:19.278 Device: null
2023-11-03 17:59:19.654
2023-11-03 17:59:19.894
2023-11-03 17:59:19.936 Got UDP broadcast from **<IP of my iOS device>**, message: -1;-100
2023-11-03 17:59:19.937 -1:-100

**Comment by cagnulein** [comment]
Hi @samkenw try to set the c1750 device on the treadmill companion and reboot it

Let me know

**Comment by samkenw** [comment]
Great, data is now being shown in the app! Thanks. 
However, controlling the treadmill does not seem to work. Am I missing a setting?

**Comment by cagnulein** [comment]
Great! In order to add control to it I need a screenshot of your manual view on the tablet that you can with adb via this script https://github.com/cagnulein/QZCompanionNordictrackTreadmill/blob/master/InstallPackage/adb-screenshot.bat

Don't u have a pc or a Mac?

**Comment by samkenw** [comment]
I have a Mac, suppose I can't just execute the bat. Will try adapting and let you know. Thanks

**Comment by cagnulein** [comment]
yes you need to call the adb lines

Il giorno ven 3 nov 2023 alle 18:48 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
yea, doing that but getting "connection refused"

**Comment by cagnulein** [comment]
did you enable adb on the treadmill tablet? you should also need to do a
factory reset on it in order to enable adb

Il giorno ven 3 nov 2023 alle 18:53 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
adb is enabled.. I have not done the factory reset however. I will do that and get back to you.

**Comment by cagnulein** [comment]
nope as far as i know adb is available only on wifi

Il giorno ven 3 nov 2023 alle 19:13 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Okay, so to clarify: do I need to reset to be able to adb connect even though I have already turned usb debugging on?

**Comment by samkenw** [comment]
Is there another way I can get the screenshot?

**Comment by cagnulein** [comment]
> Okay, so to clarify: do I need to reset to be able to adb connect even though I have already turned usb debugging on? 

Yes exactly, ifit closed the door on the latest releases.

You can use an apk that take a screenshot on the tablet and then upload the picture. It must be precise to the pixel

**Comment by samkenw** [comment]
Okay, I understand, thanks for clarifying! Do you need a screenshot of a specific screen?

**Comment by samkenw** [comment]
Sadly after reset still getting:

> adb connect <ip addr>
> * daemon not running; starting now at tcp:5037
> * daemon started successfully
**> failed to connect to '<ip addr>:5555': Connection refused**

**Comment by cagnulein** [comment]
> Okay, I understand, thanks for clarifying! Do you need a screenshot of a specific screen? 

Yes of the manual workout screen

**Comment by cagnulein** [comment]
> Sadly after reset still getting:
> 
> 
> 
> > adb connect 123.456.123.456
> 
> > * daemon not running; starting now at tcp:5037
> 
> > * daemon started successfully
> 
> **> failed to connect to '123.456.123.456:5555': Connection refused**

Ok if you want let's try a thing: clicking on the plus and minus button on the inclination tile on qz, staying on the manual workout screen on the treadmill, check you see some touch event on the tablet.
If so we have hopes!

Let me know

**Comment by samkenw** [comment]
(after reset) reading is still working but sadly no touch events visible. (I set touch to be visible in developer options)

**Comment by cagnulein** [comment]
Mmm another test
From a terminal app on the tablet try to run
`input swipe 100 200 300 400 200`

Check if you see a touch event 

Now I'm going to bed, we will catch up tomorrow 
Thanks

**Comment by samkenw** [comment]
Will try that! Can you recommend an app and mirror?

**Comment by cagnulein** [comment]
nope but there are dozens of terminal app :)

**Comment by samkenw** [comment]
Alright thanks. Btw. if you are basically simulating pressing buttons on the screen, I can give you the location of all UI elements by activating the "show pointer location" in the developer options. Additionally I can send you a photo of the screen so you know where which action is. Would this be sufficient?

**Comment by cagnulein** [comment]
yes it should be enough 4 points

max speed, min speed, max inclination, min inclinatio

then i derive the linear formula ;)

**Comment by samkenw** [comment]
sorry, what do you mean with "max speed" etc. The buttons that increase the speed on the screen for example?

**Comment by cagnulein** [comment]
the slider that you have in the manual workout window on the ifit tablet.
do you have 2 slider, right?

**Comment by samkenw** [comment]
no, I have included a picture of the screen. 
The settings are changed with manual buttons, I guess the same as on the x32i models. 
![tempImagelcJFmi](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/49457066/bc6ad66c-374b-472e-88cd-20eaa8de21c6)

**Comment by samkenw** [comment]
so sorry, I can enable the sliders. Give me a few minutes, I will pass the values

**Comment by samkenw** [comment]
Alright: 
max incline (40%) x: 75, y: 275
min incline: (-6%) x: 75, y: 845
max speed (22kph) x: 1845, y: 275
min speed (1kph): x: 1845, y: 275

**Comment by cagnulein** [comment]
Thanks, I will add them asapInviato da Apple WatchIl giorno 4 nov 2023, alle ore 18:37, Sam K ***@***.***> ha scritto:
Alright:
max incline (40%) x: 75, y: 275
min incline: (-6%) x: 75, y: 845
max speed (22kph) x: 1845, y: 275
min speed (1kph): x: 1845, y: 275

**Comment by samkenw** [comment]
Quick update: I have updated to the newest app version. Now the y-values have slightly changed (same for incline and speed) x-values are the same: 

y-low: 785 
y-high: 265

**Comment by cagnulein** [comment]
Thanks for the update!

Il giorno sab 4 nov 2023 alle 19:15 Sam K ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
done @samkenw download the new one https://github.com/cagnulein/QZCompanionNordictrackTreadmill/releases/download/3.3.13/QZCompanionNordictrackTreadmill.apk and select the new device from the list and restart the treadmill

Let me know!

**Comment by samkenw** [comment]
Thanks for the update! 

I have selected x22i in the companion app and restarted. Not getting any data in the iOS app and no control. 
Log dump: 

2023-11-05 11:25:31.41 Service started
2023-11-05 11:25:31.412 Service onCreate
2023-11-05 11:25:31.412 Service postDelayed
2023-11-05 11:25:31.412 Service started
2023-11-05 11:25:31.463 Waiting for UDP broadcast
2023-11-05 11:25:31.915 Service run
2023-11-05 11:25:31.923 /scard/.wolflogs/
2023-11-05 11:25:31.923 lastModifiedFile /sdcard/.wolflogs/2023-11-05_logs.txt
2023-11-05 11:25:31.924 string: /sdcard/.wolflogs/2023-11-05_logs.txt
2023-11-05 11:25:31.924 Parsing /sdcard/.wolflogs/2023-11-05_logs.txt
2023-11-05 11:25:31.925 Device: x221
2023-11-05 11:25:32.355 IOException: sendto failed: ENETUNREACH (Network is unreachable)
2023-11-05 11:25:32.442 IOException: sendto failed: ENETUNREACH (Network is unreachable)
2023-11-05 11:25:32.609 IOExcention: sendto failed: ENETUNREACH (Network is unreachable)

I have also tested with selecting c1750: this is still transmitting data to the app but no control.

**Comment by cagnulein** [comment]
ok so please test the adb swipe thing that I said above. it's the only way
to override the ifit protection

Il giorno dom 5 nov 2023 alle 11:59 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Tested the sipe, I am seeing an input event. 
but as I said, when selecting x22i in the companion I am also not receiving any data

**Comment by cagnulein** [comment]
ok so it means I need to implement the adb method in qz, no problem I
already did it.

Right now this method is available for android or windows. is one of this
ok for you?

otherwise It means I have to do a bigger development

Il giorno dom 5 nov 2023 alle 12:30 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
I sadly have neither android nor windows... will the adb method work even though I cannot use adb to connect to the treadmill from my Mac?

**Comment by cagnulein** [comment]
do you have a m1-2-3 mac or an intel one?

Il giorno dom 5 nov 2023 alle 12:35 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
intel

**Comment by cagnulein** [comment]
ok i can try a build but only starting from wednesday

also using the mac version will require to become a patreon member due to
the high development required

Il giorno dom 5 nov 2023 alle 12:37 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
tell you what, I have a raspberry on which I can try running adb - would this help?

**Comment by samkenw** [comment]
I fully understand the requirement for patreon. However this is currently outside of my scope, so perhaps don't develop this just for me. I will try a VM for windows.

**Comment by cagnulein** [comment]
also the window version requires the patreon one but at least it's already
available:)

Il giorno dom 5 nov 2023 alle 12:41 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
I understand. I am not certain that adb will work for this model as it is running android 5.1.1 and it seems that network debugging may not be possible anyway.

**Comment by cagnulein** [comment]
how did you run the adb swipe thing? from terminal?

if so can you please dump log from the companion just after pressing the
plus button of the speed or inclination tile

maybe there is still something to fix, let me know

Il giorno dom 5 nov 2023 alle 12:47 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Yes from a terminal.
Perhaps this is the issue: it sees and transmits from treadmill to app correctly - but values received from app are invalid

Log dump:

2023-11-05 12:52:59.668 Parsing /sdcard/wolflogs/2023-11-05_logs.txt
2023-11-05 12:52:59.674 Device: c1750
2023-11-05 12:52:59.9 [35] **3360145 12:52:44.4274 [Trace:FitPro] Changed KPH to: 3**
2023-11-05 12:52:59.905 [27] 3360438 12:52:44.7205 (Trace:FitPro] Changed Watts to: 119
2023-11-05 12:53:00.028 **Got UDP broadcast from 192.168.178.21, message: -1;-100**
2023-11-05 12:53:00.031 -1;-100
2023-11-05 12:53:00.035 **requestSpeed: -1.0 lastSpeed:3.0 cachedSpeed:-1.0**
2023-11-05 12:53:00.038 **requestInclination: -100.0 -100.0**
2023-11-05 12:53:00.046 Waiting for UDP broadcast
2023-11-05 12:53:00.05 Waiting for UDP broadcast

**Comment by samkenw** [comment]
In the app, I have no treadmill selected - is there a specific option I need to set for the correct values to be transmitted?

**Comment by cagnulein** [comment]
> In the app, I have no treadmill selected - is there a specific option I need to set for the correct values to be transmitted? 
> 
> 

that's right

**Comment by samkenw** [comment]
Which treadmill would I need to select? The x22i is not an option.

**Comment by cagnulein** [comment]
in the log that you sent i don't see the requested speed, probably you got a moment after the send (qz is always send that formula so you can scroll the dump log looking for a speed request with a value different from -1)

**Comment by cagnulein** [comment]
> Which treadmill would I need to select? The x22i is not an option. 

the ip is enough, the device is selected in the companion

**Comment by samkenw** [comment]
Is this not the message received from the app? Seems like the app is broadcasting the wrong values?

2023-11-05 12:53:00.028 Got UDP broadcast from 192.168.178.21, message: -1;-100
2023-11-05 12:53:00.031 -1;-100
2023-11-05 12:53:00.035 requestSpeed: -1.0 lastSpeed:3.0 cachedSpeed:-1.0
2023-11-05 12:53:00.038 requestInclination: -100.0 -100.0

**Comment by cagnulein** [comment]
as i said qz is always sending those values. when it has a request from the
user it change -1 to the value that you request.

so you have to capture the right moment in the log or simply send me logcat
file available in the root of the tablet so i can check the whole file
myself

Let me know

Il giorno dom 5 nov 2023 alle 13:05 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Ah I understand, sorry.

Here is a request: 
2023-11-05 13:16:11.932 (43] 4126844 13:05:31.1260 [Trace:FitPro) Changed Watts to: 0
2023-11-05 13:16:12.036 Got UDP broadcast from 192.168.178.21, message: 0.5;-100
2023-11-05 13:16:12.036 0.5;-100
2023-11-05 13:16:12.037 requestSpeed: 0.5 lastSpeed:0.0 cachedSpeed:-1.0
2023-11-05 13:16:12.037 requestinclination: -100.0 -100.0
2023-11-05 13:16:12.04 Waiting for UDP broadcast
2023-11-05 13:16:12.043 Got UDP broadcast from 192.168.178.21, message: -1;-100
2023-11-05 13:16:12.044 -1;-100
2023-11-05 13:16:12.044 requestSpeed: -1.0 lastSpeed:0.0 cachedSpeed:0.5
2023-11-05 13:16:12.045 requestInclination: -100.0 -100.0

**Comment by cagnulein** [comment]
perfect! the reason why it's not working it's because qz companion wants to
see a speed greater than 0 to send the touch event.

try to start the treadmill to a speed of 1 for example, wait qz to shows
it, and then try to change it

let me know

Il giorno dom 5 nov 2023 alle 13:19 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Alright, now with a running treadmill and set to c1750:

2023-11-05 12:35:23.727 Parsing /sdcard/wolflogs/2023-11-05_logs.txt
2023-11-05 12:35:23.729 Device: c1750
2023-11-05 12:35:24.009 (7] 379869 12:34:54.4954 [Trace:DemoModeServiceSelector] Changed KPH to: 2
2023-11-05 12:35:24.011 [7] 197746 12:31:52.3710 [Trace:RetailDemoMode1ServiceAdapter] Changed Watts to: 107
2023-11-05 12:35:24.045 Got UDP broadcast from 192.168.178.21, message: 2.5;-100
2023-11-05 12:35:24.046 2.5;-100
2023-11-05 12:35:24.046 requestSpeed: 2.5 lastSpeed:2.0 cachedSpeed:-1.0
2023-11-05 12:35:24.048 input swipe 1845 690 1845 706 200
2023-11-05 12:35:24.051 Waiting for UDP broadcast
2023-11-05 12:35:24.514 Service run
2023-11-05 12:35:24.523 /sdcard/wolflogs/
2023-11-05 12:35:24.525 lastModifiedFile /sdcard/wolflogs/2023-11-05_logs.txt

**Comment by cagnulein** [comment]
perfect now but you have to use the x22i device in the companion that i
created for you and remember to reboot the treadmill

Il giorno dom 5 nov 2023 alle 13:33 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
I will try but when selecting the x22i there was no feedback in the QZ app. I will try again and let you know.

**Comment by cagnulein** [comment]
argh you are right! i missed a thing! i will do later today

sorry

Il giorno dom 5 nov 2023 alle 13:42 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
okay great, thanks!

**Comment by cagnulein** [comment]
done https://github.com/cagnulein/QZCompanionNordictrackTreadmill/releases/download/3.3.14/QZCompanionNordictrackTreadmill.apk sorry for the issue

**Comment by samkenw** [comment]
Great! Now seeing the changes in the QZ app. However, no input events seen when changing the values.

Log Dump:
2023-11-05 16:19:42.003 Parsing /scard/wolflogs/2023-11-05 _logs.txt
2023-11-05 16:19:42.014 Device: x221
2023-11-05 16:19:42.396 [20] 365957 16:17:55.8296 [Trace:FitPro] Changed KPH to: 2
2023-11-05 16:19:42.399 [10] 366875 16:17:56.7470 [Trace:FitPro] Changed Watts to: 107
2023-11-05 16:19:42.438 Got UDP broadcast from 192.168.178.21, message: -1;0.5
2023-11-05 16:19:42.44-1:0.5
2023-11-05 16:19:42.442 requestSpeed: -1.0 lastSpeed:2.0 cachedSpeed:-1.0
2023-11-05 16:19:42.446 requestinclination: 0.5 - 100.0
2023-11-05 16:19:42.449 input swipe 75 785 75 711 200
2023-11-05 16:19:42.454 Waiting for UDP broadcast
2023-11-05 16:19:42.456 Waiting for UDP broadcast
2023-11-05 16:19:42.905 Service run
2023-11-05 16:19:42.918 /sdcard/wolflogs/
2023-11-05 16:19:42.923 lastModifiedFile /sdcard/wolflogs/2023-11-05_logs.txt

**Comment by cagnulein** [comment]
input swipe are now sent correctly. can you check if you can set qz
companion as an adb debug app inn the developer setting on the treadmill
tablet?

Il giorno dom 5 nov 2023 alle 16:27 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Just tried. Also restart treadmill afterwards. Sadly no success. 
Can this maybe be the same issue that I can't connect via adb over wifi?

**Comment by samkenw** [comment]
another thing: each time I restart the treadmill, I have to go back into developer options and enable usb debugging... not sure why this is being turned off again.. perhaps there is some lock on it? (I am on the newest software version)

**Comment by cagnulein** [comment]
nope it shouldn't be about this because you can with another terminal app.
I can try another approach tomorrow
We are very near!

Il giorno dom 5 nov 2023 alle 16:40 Sam K ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
> another thing: each time I restart the treadmill, I have to go back into developer options and enable usb debugging... not sure why this is being turned off again.. perhaps there is some lock on it? (I am on the newest software version) 

yes this is normal on the newer ifit version

**Comment by samkenw** [comment]
alright, thanks for the effort up to now !

**Comment by samkenw** [comment]
So, 2 things I noticed yesterday:

1. It seems that the companion app stops transferring data after a certain point. To fix this, I had to switch to the companion app (bring it to the foreground) and then switch back. Then it was fine again for a while. Could it be that the app is being put to sleep after a while? 

2. I was not getting any heart rate data from the treadmill - is there an easy fix for this?

**Comment by cagnulein** [comment]
> 1. It seems that the companion app stops transferring data after a certain point. To fix this, I had to switch to the companion app (bring it to the foreground) and then switch back. Then it was fine again for a while. Could it be that the app is being put to sleep after a while?

mmm interesting, I never noticed this on others. I can put a notification on the companion so the app will be forced to stay awake. I'm doing the mod today along with the new input swipe thing 


> 2\. I was not getting any heart rate data from the treadmill - is there an easy fix for this?

can you open the file /sdcard/wolflogs/2023-11-05 _logs.txt and check if you see the HR there? if so I can parse it for sure!

**Comment by samkenw** [comment]
Great and great, will check the logs and get back to you!

**Comment by cagnulein** [comment]
ok i did the input swipe patch. let me know if it's better https://github.com/cagnulein/QZCompanionNordictrackTreadmill/releases/download/3.3.15/QZCompanionNordictrackTreadmill.apk

**Comment by samkenw** [comment]
Great, will test it shortly!

This is the line for the heart rate from the logs: 

<timestamp etc> [Trace: HeartRateConnection] HeartRateConnection. HeartRateDataUpdate heart rate value changed. Value: 127 Device: <device id>

**Comment by cagnulein** [comment]
perfect! added! for HR you need to update both QZ companion to the 3.3.16 and QZ to the last build, for this 
send me an email to roberto.viola83 at gmail.com indicating this ticket and if you are on android or iOS. Thanks

**Comment by samkenw** [comment]
Awesome, thanks! will do!

Sadly, no change on the swipe issue:



  | 2023-11-06 09:19:41.295 [11) 201742 09:19:24.0759 (Trace:FitPro) Changed KPH to: 2
-- | --
  | 2023-11-06 09:19:41.299 41 202849 09:19:25.1835 [Trace:FitPro, Changed Watts to: 107
  | 2023-11-06 09:19:41.41 Got UDP broadcast from _ip address_, message: 2.5;100
2023-11-06 09:19:41.416 2.5;-100 |  
  | 2023-11-06 09:19:41.424 requestSpeed: 2.5 lastSpeed:2.0 cachedSpeed:-1.0
  | 2023-11-06 09:19:41.433 requestSpeed: -1.0 lastSpeed:2.0 cachedSpeed:-1.0

**Comment by cagnulein** [comment]
new companion for the HR https://github.com/cagnulein/QZCompanionNordictrackTreadmill/releases/download/3.3.16/QZCompanionNordictrackTreadmill.apk

I don't see the input swipe line in your log, could you please post the full log?

**Comment by samkenw** [comment]
Thanks! 
There is no further output. If I see correctly, in the shellRuntime.exec there is no logging though, is that correct?

**Comment by cagnulein** [comment]
yes but on line 280 there is a writeLog(command);

**Comment by samkenw** [comment]
My bad, updated previous comment.

**Comment by cagnulein** [comment]
ok so there is the input swipe. 
could you run that exactly line in the terminal to check that you can see the swipe from the terminal?

**Comment by samkenw** [comment]
This is returning "killed" in the terminal

**Comment by cagnulein** [comment]
interesting, so what's the difference from the one that worked before?

**Comment by samkenw** [comment]
trying to figure that out.. the other one still works

**Comment by cagnulein** [comment]
can you put here the one that it's working?

also try to change one by one the parameters so we will quickly understand which is the issue :)

We're so near!

**Comment by samkenw** [comment]
input swipe 1845 200 1845 400 200 works 
input swipe 1845 725 1845 525 200 fails :D 

will do some testing and get back

**Comment by samkenw** [comment]
Alright so: 
**input swipe 1845 650 1845 650 200 works**
**input swipe 1845 680 1845 680 200 fails** - but this is exactly where the onscreen keyboard starts. Could it be that there is some lock preventing inputs on specific components? Perhaps the treadmill app is also blocking programmatic inputs to the application ?

**Comment by cagnulein** [comment]
> Perhaps the treadmill app is also blocking programmatic inputs to the application ?

never saw this before so I don't think so

> but this is exactly where the onscreen keyboard starts

on screen keyboard? what do you mean? aren't we sliding the speed and incline slider with these coordinates?

Also I was checking again your log and 

`input swipe 1845 725 1845 725 200`

and the y0 and y1 are the same! so that's why you don't see anything. I'm checking the code again

**Comment by cagnulein** [comment]
it seems that in that line, the companion already did that swipe, the first "725" was the last y1 coordinate that QZ sends. Try for example to send

`input swipe 1845 785 1845 725 200`

on the terminal app

**Comment by samkenw** [comment]
The onscreen keyboard in the terminal application I mean. That is how I am testing the swipes. As soon as the coordinates are on the keyboard, the command fails. 

Same y positions is not an issue. When doing this in the terminal application I can see that there is an interaction.

**Comment by samkenw** [comment]
input swipe 1845 785 1845 725 200 **fails** because the coordinates hit the keyboard
input swipe 1845 585 1845 525 200 **works**

**Comment by cagnulein** [comment]
ok so you can test from the terminal app to ifit for example doing

`sleep 10 && input swipe 1845 725 1845 525 200 `

in the sleep time, go to the ifit app 

do you know what I mean?

**Comment by samkenw** [comment]
Tested this. Even with values that work in the terminal app, when switching I am getting "Killed" return. 
Looks like cross-process input is somehow being blocked?

**Comment by cagnulein** [comment]
mmm yes I guess so. So we have 2 possibilites now
1) trying to reset the console, don't update ifit and see if you have the same behaviour
2a) if 1. is working, keep the old ifit
2b) if 2. is not working, we need to use the adb method somehow

Let me know

**Comment by samkenw** [comment]
okay, I can test that later some time. 
However, could it also be possible that there is a permission missing in the manifest? 
something like: **uses-permission android:name="android.permission.INJECT_EVENTS"** ?

**Comment by cagnulein** [comment]
> Ah but I am seeing that for this to work the app needs to be signed by the system.

yes but I also found this on the same thread https://stackoverflow.com/a/25014038

Can you access to that XML file?

**Comment by cagnulein** [comment]
also check this https://www.pocketmagic.net/injecting-events-programatically-on-android/
there is an apk to test it at the bottom

**Comment by samkenw** [comment]
Access to the XML privileges that I don't have.. (I believe the APK also needs a rooted device)

**Comment by amasolov** [comment]
@cagnulein experience the same problem on my NT x22i, the latest build of the companion and the last version of QZ (tried the google play version and the latest debug one from the repository). Would be interested to help if needed and see this resolved

**Comment by cagnulein** [comment]
> Actually, I just realized that I need to enable onscreen controls (sliders) to be able to control everything and it has started to work. Can we get a new build of QZ in Google Play to avoid the trial period of the debug version?

yes just join the beta from the play store @amasolov 

@samkenw did you enable the onscreen controls that @amasolov  mentioned? I guess so

**Comment by samkenw** [comment]
@cagnulein yea, onscreen controls are active.
@amasolov which version of the ifit software are you on?

**Comment by amasolov** [comment]
@samkenw I'm on iFit 2.6.88. I installed the companion with ADB by cloning the git repository and running QZ-Companion.bat on a Windows 11 PC

**Comment by samkenw** [comment]
@amasolov thanks for the info! Can you also tell me which android your machine is running?

**Comment by samkenw** [comment]
@cagnulein finally got adb working - perhaps we can test via this again

**Comment by samkenw** [comment]
In the terminal app on the treadmill I ran: 
```
setprop service.adb.tcp.port 5555
stop adbd
start adbd
```
This enabled adb. 

I need to be able to control it from iOS, is this not possible with the previous version?

**Comment by samkenw** [comment]
I have no experience with iOS development - but if there is any info on what to do I can try and help :)

**Comment by cagnulein** [comment]
ok i will try a quick test later this week. if we will lucky maybe it's
already everything developed

Il giorno lun 6 nov 2023 alle 16:48 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
awesome! :)

**Comment by cagnulein** [comment]
ok @samkenw  I could be ready for a test but I can't test it myself, so It will be good if you could be able to run a simple xcode project from your pc to the treadmill.

If I push the project here on github, can you run xcode and test it?

**Comment by samkenw** [comment]
@cagnulein yea definitely, just tell me how and I will test it later today! :)

**Comment by cagnulein** [comment]
@samkenw perfect! let's start from this: close this repository https://github.com/cagnulein/adb-ios and try to open and build with xcode. Let me know if you can without issues

**Comment by samkenw** [comment]
Build was successful with some warnings.
Btw. Have you checked this: [https://stackoverflow.com/questions/28337921/execute-adb-shell-input-swipe-command-from-apk](url)

**Comment by cagnulein** [comment]
it's what I'm already do but you don't have root permissions on your bike
I will give you the further instructions in a couple of hours
thanks

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 8 nov 2023 alle ore 13:33 Sam K ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
ok now @samkenw in the viewcontroller.m change the IP

`#define IP  "192.168.0.46:43832"`

with the IP of your tablet

`#define IP  "192.168.0.xx"`

and start it on your mac and try to connect to it

**Comment by samkenw** [comment]
Okay I will try tomorrow. But for some reason Xcode is not giving me the option to actually run the app... only build. Do you know what I need to change?

Btw. I am seeing in the Logs that the treadmill is also outputting "Actual Speed" which is slightly different than the speed that is set - would this be worth considering using instead of what the user inputs? I believe that the "actual speed" could be the speed reading from the speed sensor. (no idea why this is not the speed displayed in the console)

**Comment by amasolov** [comment]
@cagnulein I have tried it on mine as well. The build went fine. This is what I see in the console:

```
/Users/amasolov/projects/adb-ios/adb/adb/adb_client.c::adb_query():adb_query: host:connect:192.168.2.233
/Users/amasolov/projects/adb-ios/adb/adb/adb_client.c::_adb_connect():_adb_connect: host:version
/Users/amasolov/projects/adb-ios/adb/adb/adb_client.c::adb_connect():adb_connect: service host:connect:192.168.2.233
* daemon not running. starting it now on port 5037 *
* daemon started successfully *
/Users/amasolov/projects/adb-ios/adb/adb/transport_local.c::local_init():transport: local client init
/Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::adb_auth_init():adb_auth_init
/Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::get_user_keyfilepath():home '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app'
/Users/amasolov/projects/adb-ios/adb/adb/transport_local.c::client_socket_thread():transport: client_socket_thread() starting
/Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::get_user_key():user key '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app/android/adbkey'
/Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::read_key():read_key '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app/android/adbkey'
LSPrefs: could not find untranslocated node for <FSNode 0x600003fe83a0> { isDir = ?, path = '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app' }, proceeding on the assumption it is not translocated: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted
``` 

@samkenw It could be that you don't have a simulator installed to run the build.

**Comment by cagnulein** [comment]
> Okay I will try tomorrow. But for some reason Xcode is not giving me the option to actually run the app... only build. Do you know what I need to change?

check the @amasolov answer about this


> Btw. I am seeing in the Logs that the treadmill is also outputting "Actual Speed" which is slightly different than the speed that is set - would this be worth considering using instead of what the user inputs? I believe that the "actual speed" could be the speed reading from the speed sensor. (no idea why this is not the speed displayed in the console) 

yes sure but I guess we need to create a setting for this because most of the people don't want to see a value that it's different from the one that set

**Comment by cagnulein** [comment]
> @cagnulein I have tried it on mine as well. The build went fine. This is what I see in the console:
> 
> 
> 
> ```
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_client.c::adb_query():adb_query: host:connect:192.168.2.233
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_client.c::_adb_connect():_adb_connect: host:version
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_client.c::adb_connect():adb_connect: service host:connect:192.168.2.233
> 
> * daemon not running. starting it now on port 5037 *
> 
> * daemon started successfully *
> 
> /Users/amasolov/projects/adb-ios/adb/adb/transport_local.c::local_init():transport: local client init
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::adb_auth_init():adb_auth_init
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::get_user_keyfilepath():home '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app'
> 
> /Users/amasolov/projects/adb-ios/adb/adb/transport_local.c::client_socket_thread():transport: client_socket_thread() starting
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::get_user_key():user key '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app/android/adbkey'
> 
> /Users/amasolov/projects/adb-ios/adb/adb/adb_auth_host.c::read_key():read_key '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app/android/adbkey'
> 
> LSPrefs: could not find untranslocated node for <FSNode 0x600003fe83a0> { isDir = ?, path = '/private/var/folders/8t/c5y7cdfj7_n31dsnfpjt43_h0000gn/X/8B97A60B-1271-57AD-902F-2E21B9500709/d/Wrapper/adb-ios.app' }, proceeding on the assumption it is not translocated: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted

ok it's the same thing that I saw on my android 14 devices. now try to press the list packages button, i did a change from the original code and that button actually simple invoke a "ls" on adb

let me know @amasolov

**Comment by amasolov** [comment]
@cagnulein Nothing happens when I choose the list of packages - nothing in the debug console and nothing in the app

**Comment by cagnulein** [comment]
put a break in the ls line it will show you the output

on android 14 it says device offline

Il giorno gio 9 nov 2023 alle 06:46 Alexey Masolov ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
![image](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/48216/957124c9-42d1-4384-8b7a-baeda8027c6b)

here in the line 91

**Comment by amasolov** [comment]
yup, I changed it to "ls\n" but no response

**Comment by amasolov** [comment]
`result	__NSCFString *	@"acct       data             init.mt8163.usb.rc      lost+found sdcard     \nbin        default.prop     init.rc                 mnt        storage    \nbugreports dev              init.recovery.mt8163.rc odm        sys        \ncache      etc              init.usb.configfs.rc    oem        system     \ncharger    fstab.enableswap init.usb.rc             proc       ueventd.rc \nconfig     init             init.zygote32.rc        product    vendor     \n"	`

**Comment by amasolov** [comment]
@cagnulein Yes, we are getting the list of files on the tablet

**Comment by cagnulein** [comment]
amazing so it's working!

great! i will implement in qz today!!

**Comment by cagnulein** [comment]
ok guys implemented! 
send me an email to roberto.viola83 at gmail.com indicating this ticket and I will send you the new beta today!

I did not add the right coordinates for your bike, but I would like just to see if you see a reaction when you change the speed and incline on QZ on iOS to the touchscreen of the treadmill.

So in order to enable the patch, enable the adb remote setting under the norditrack treadmill and restart QZ.
It's of course highly in alpha version, so enable the debug log and send me it in any case so I can check it

Thanks!

**Comment by cagnulein** [comment]
@victorypoint FYI I also added the adb bridge to iOS!

**Comment by victorypoint** [comment]
> @victorypoint FYI I also added the adb bridge to iOS!

This is great news! Nice work guys.

**Comment by amasolov** [comment]
@cagnulein The QZ beta shows zero values in the tiles, unfortunately :( It's with or without ADB remote enabled. The android version still works

**Comment by cagnulein** [comment]
@amasolov that's not about the today's beta @samkenw can read the metrics even with the previous build.

let's upload and compare the log from android to the ios one

**Comment by cagnulein** [comment]
any news guys?

**Comment by samkenw** [comment]
Sent you an email but have not received an iOS app to test - could you check?

**Comment by cagnulein** [comment]
@samkenw check the spam folder from apple, i sent it yesterday along with @amasolov 

![image](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/48216/629d044c-a05c-4b1d-a459-80f8fc7ed06c)

**Comment by samkenw** [comment]
Okay I will check thanks. Could you please remove my email address from your post - this is private.

**Comment by cagnulein** [comment]
done, sorry

**Comment by samkenw** [comment]
No worries, thanks :) received the invite, will test later!

**Comment by amasolov** [comment]
Sorry, busy day today at work. Will test more on the weekend!

**Comment by cagnulein** [comment]
thanks, no rush i was only curious to fix it :)

Il giorno ven 10 nov 2023 alle 10:42 Alexey Masolov <
***@***.***> ha scritto:

**Comment by samkenw** [comment]
Alright, tested the app.
With adb remote on and off same result: touches are visible when the companion app is in the foreground. However, no touches visible in ifit or anywhere else.

**Comment by cagnulein** [comment]
@samkenw please attach a debug log with adb remote enabled 

also be sure you are still able to use adb from your mac to send events to the ifit app.

**Comment by samkenw** [comment]
I am not seeing any logs from the companion at: /sdcard/logcat.log - is there an alternate location?

**Comment by cagnulein** [comment]
i mean the debug log from the qz main app

Il giorno ven 10 nov 2023 alle 17:09 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
where can I find these?

**Comment by samkenw** [comment]
Sent per email.

**Comment by cagnulein** [comment]
answered, also what about the adb on mac test?

Il giorno ven 10 nov 2023 alle 17:50 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
adb on Mac works as expected (on ifit app)

**Comment by cagnulein** [comment]
@samkenw i'm an idiot I did the modification on the nordictrack bike! i'm doing the same on the treadmill and i will send you it again

There is still hope!

**Comment by cagnulein** [comment]
@amasolov @samkenw new version sent! let me know!

**Comment by samkenw** [comment]
Getting this in the logs:

Fr. Nov. 10 19:42:20 2023 1699641740426 Debug: /Users/cagnulein/qdomyos-zwift/src/ios/lockscreen.mm void lockscreen::adb_sendcommand(const char *)_block_invoke device unauthorized. Please check the confirmation dialog on your device.

But its not showing a dialog.

**Comment by cagnulein** [comment]
@samkenw maybe we can inject the permissions for the iphone ip somehow in the tablet?

I have to think about this

**Comment by samkenw** [comment]
It worked! But only VERY randomly and only in one position on the screen. I pressed buttons randomly and sometimes it gave an input. Will send you the logs.

**Comment by samkenw** [comment]
But apparently it only works in the companion app ... and only in one position and very randomly.

**Comment by cagnulein** [comment]
wow! amazing! yes i didn't set the coordinates yet! so this is a success!!

what did you do for the authorization issue?

i will add the coordinates in the next days!

Il giorno ven 10 nov 2023 alle 20:21 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
I did nothing for the auth issue. But as I said: I smash all of the buttons and at some point there is suddenly an event - but only inside the companion app so far - not in ifit, browser, desktop...  I have no reliable way to reproduce it reliably.

**Comment by cagnulein** [comment]
checking the log i guess the companion is still receiving the old udp
commands, that's why.

we need to understand the auth issue

Il giorno ven 10 nov 2023 alle 20:34 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Hm okay.. I suppose you are doing a standard `adb connect` first?

**Comment by samkenw** [comment]
```
Fr. Nov. 10 20:15:28 2023 1699643728708 Debug: /Users/cagnulein/qdomyos-zwift/src/ios/lockscreen.mm void lockscreen::adb_connect(const char *)_block_invoke connected to 192.168.178.42:5555
Fr. Nov. 10 20:15:28 2023 1699643728715 Debug: /Users/cagnulein/qdomyos-zwift/src/ios/lockscreen.mm void lockscreen::adb_sendcommand(const char *)_block_invoke device offline
```

I guess this is perhaps part of the issue?

Would it be useful to connect the phone to the tablet via USB? (Mine has a USB port that recognizes the phone)

**Comment by SOLLY9** [comment]
I connected my phone to the bike usb. Now what should I do?

**Comment by cagnulein** [comment]
> ```
> 
> Fr. Nov. 10 20:15:28 2023 1699643728708 Debug: /Users/cagnulein/qdomyos-zwift/src/ios/lockscreen.mm void lockscreen::adb_connect(const char *)_block_invoke connected to 192.168.178.42:5555
> 
> Fr. Nov. 10 20:15:28 2023 1699643728715 Debug: /Users/cagnulein/qdomyos-zwift/src/ios/lockscreen.mm void lockscreen::adb_sendcommand(const char *)_block_invoke device offline
> 
> ```
> 
> 
> 
> I guess this is perhaps part of the issue?
> 
> 
> 
> Would it be useful to connect the phone to the tablet via USB? (Mine has a USB port that recognizes the phone) 

yes for sure it's the issue. nope usb it's not handled on ios and it can't be. did you have the same issue with the xcode project that i sent you?

**Comment by cagnulein** [comment]
> I connected my phone to the bike usb. Now what should I do?
> 
> 

so adb is working now on your bike from your pc? also wifi adb?

**Comment by SOLLY9** [comment]
Hi again. I was lost trying to connect via adb. I tried the script in a terminal emulator on my bike which is shown above but it did not work. I still cannot connect via adb but did get somewhere by resetting the bike a couple of times and then reinstalling qz companion by using the browser. I detailed the results of this on my open report.

**Comment by samkenw** [comment]
> yes for sure it's the issue. nope usb it's not handled on ios and it can't be. did you have the same issue with the xcode project that i sent you?

I can't get it to run in the simulator - which version of iOS is required?

**Comment by samkenw** [comment]
> ```
> setprop service.adb.tcp.port 5555
> stop adbd
> start adbd
> ```

@cagnulein Just did this again, now the dialog appeared and it works (from the beta app on iOS) !!! Seeing touch events even on the ifit app!

**Comment by samkenw** [comment]
Btw. incase you need to correct the pixel values:
![now](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/49457066/a12ed878-fe3d-447e-8dab-89952b1dd120)

**Comment by cagnulein** [comment]
@samkenw amazing! i will add your coordinates in the next one. in the weekend i don't have much time since i would like to stay with my family, hopefully on monday! i will keep you updated

**Comment by samkenw** [comment]
No rush! Would be awesome if we can also test HR from the machine to QZ.

**Comment by cagnulein** [comment]
the hr mod is already in the build that you have! give it a try , remember
to update the companion as well if you didn't

Il giorno sab 11 nov 2023 alle 10:47 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
It looks like the HR data is arriving in the App, however it is not being displayed. 

`Sa. Nov. 11 18:19:23 2023 1699723163666 Debug: /Users/cagnulein/qdomyos-zwift/src/nordictrackifitadbtreadmill.cpp void nordictrackifitadbtreadmill::processPendingDatagrams() Message From ::  "ip"
Sa. Nov. 11 18:19:23 2023 1699723163666 Debug: /Users/cagnulein/qdomyos-zwift/src/nordictrackifitadbtreadmill.cpp void nordictrackifitadbtreadmill::processPendingDatagrams() Port From ::  53230
Sa. Nov. 11 18:19:23 2023 1699723163666 Debug: /Users/cagnulein/qdomyos-zwift/src/nordictrackifitadbtreadmill.cpp void nordictrackifitadbtreadmill::processPendingDatagrams() Message ::  "[65] 8481605 18:19:23.8277 [Trace:HeartRateConnection] HeartRateConnection.HeartRateDataUpdate heart rate value changed. Value: 157 Device: id
Sa. Nov. 11 18:19:23 2023 1699723163667 Debug: /Users/cagnulein/qdomyos-zwift/src/nordictrackifitadbtreadmill.cpp void nordictrackifitadbtreadmill::processPendingDatagrams() "[65] 8481605 18:19:23.8277 [Trace:HeartRateConnection] HeartRateConnection.HeartRateDataUpdate heart rate value changed. Value: 157 Device: id"
Sa. Nov. 11 18:19:23 2023 1699723163667 Debug: /Users/cagnulein/qdomyos-zwift/src/nordictrackifitadbtreadmill.cpp void nordictrackifitadbtreadmill::processPendingDatagrams() "7 >> -1;-100"

Sa. Nov. 11 18:19:23 2023 1699723163670 Debug: /Users/cagnulein/qdomyos-zwift/src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"`

**Comment by cagnulein** [comment]
heart rate belt is set to disabled and also disabled hr from machinery is
disabled?

Il giorno sab 11 nov 2023 alle 19:06 Sam K ***@***.***> ha
scritto:

**Comment by amasolov** [comment]
@cagnulein I'm still getting zero values in QZ TestFlight. Just sent you my debug log file by email

**Comment by cagnulein** [comment]
> @cagnulein I'm still getting zero values in QZ TestFlight. Just sent you my debug log file by email

@amasolov log checked, the iOS version is not getting any frame from the companion. If the android one works instead, it means that probably the wifi is on a different channel or a different wifi?

The companion is sending the data in broadcasting so it doesn't make difference between iOS or Android.

**Comment by amasolov** [comment]
Actually, I just fixed it. I'm not sure what is the root cause but it has something to do with having my treadmill and my phone into separate network subnets. Wondering if ADB doesn't work in a routed environment. Will have a read about it and report back

**Comment by cagnulein** [comment]
> I'm not sure what is the root cause but it has something to do with having my treadmill and my phone into separate network subnets

hah ok, it makes sense.



> Wondering if ADB doesn't work in a routed environment. Will have a read about it and report back

ok let me know!

**Comment by cagnulein** [comment]
@samkenw I'm building the version, when you will receive it, remember to enable the x22i setting in QZ as well and restart QZ. Let me know when you will try it!

**Comment by samkenw** [comment]
Awesome! Will test it asap. 
(btw. in the app settings I did have a chest strap device set for HR so perhaps that is why it wasn't showing - will check this also and let you know)

**Comment by cagnulein** [comment]
@samkenw @amasolov new beta sent to both, enable the x22i setting in the nordictrack settings in qz main app on ios

**Comment by samkenw** [comment]
Awesome! Right, so adjusting incline and speed is working great! (its not very accurate - sometimes going up by 0,1 and sometimes by 0,5 but I expect that is because of the slider not the logic - but I think that's fine) 

However, heart rate is not coming through.. (I am not using an Apple Watch)

`Mo. Nov. 13 17:34:41 2023 1699893281956 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "ant_heart" QVariant(bool, false)
Mo. Nov. 13 17:34:41 2023 1699893281958 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "bike_heartrate_service" QVariant(bool, false)
Mo. Nov. 13 17:34:41 2023 1699893281976 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "heart_ignore_builtin" QVariant(bool, false)
Mo. Nov. 13 17:34:41 2023 1699893281976 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "heart_max_override_enable" QVariant(bool, false)
Mo. Nov. 13 17:34:41 2023 1699893281976 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "heart_max_override_value" QVariant(double, 195)
Mo. Nov. 13 17:34:41 2023 1699893281977 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "heart_rate_belt_name" QVariant(QString, "Disabled")
Mo. Nov. 13 17:34:41 2023 1699893281981 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "ios_cache_heart_device" QVariant(bool, true)
Mo. Nov. 13 17:34:42 2023 1699893282071 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "treadmill_pid_heart_max" QVariant(qlonglong, 0)
Mo. Nov. 13 17:34:42 2023 1699893282071 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "treadmill_pid_heart_min" QVariant(qlonglong, 0)
Mo. Nov. 13 17:34:42 2023 1699893282071 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "treadmill_pid_heart_zone" QVariant(QString, "Disabled")
Mo. Nov. 13 17:34:42 2023 1699893282083 Debug: /Users/cagnulein/qdomyos-zwift/src/main.cpp int main(int, char **) "virtual_device_onlyheart" QVariant(bool, false)
Mo. Nov. 13 17:34:42 2023 1699893282734 Debug: /Users/cagnulein/qdomyos-zwift/src/dirconmanager.cpp DirconManager::DirconManager(bluetoothdevice *, uint8_t, double, QObject *) Initializing dircon for "Wahoo HRM" with serv "servname=HEART_RATE uuid=180d machine=1 [uuid=2a37 read=00 type=4,],"
Mo. Nov. 13 17:34:43 2023 1699893283090 Debug: /Users/cagnulein/qdomyos-zwift/src/bluetoothdevice.cpp void bluetoothdevice::update_hr_from_external() Current Heart from Apple Watch:  "0"
Mo. Nov. 13 17:34:43 2023 1699893283092 Debug: /Users/cagnulein/qdomyos-zwift/src/bluetooth.cpp void bluetooth::debug(const QString &) "Current Heart: 0"
`

**Comment by cagnulein** [comment]
> Awesome! Right, so adjusting incline and speed is working great! (its not very accurate - sometimes going up by 0,1 and sometimes by 0,5 but I expect that is because of the slider not the logic - but I think that's fine)

wow amazing! actually it should be linear, maybe we just need to find the right value.
If you want you can experiment with the coordinates if you have  time: the parameter passed to adb are the following https://github.com/cagnulein/qdomyos-zwift/commit/c0bba3943b71c36c6574cc1c3fef4baebf9791c2

> However, heart rate is not coming through.. (I am not using an Apple Watch)

please attach here the full log, thanks

**Comment by cagnulein** [comment]
@samkenw in the log that you sent me I can't see the HR from the companion, did you see before?

**Comment by samkenw** [comment]
Yea, I did see it before.. 

Sa. Nov. 11 18:05:22 2023 1699722322230 Debug: /Users/cagnulein/qdomyos-zwift/src/nordictrackifitadbtreadmill.cpp void nordictrackifitadbtreadmill::processPendingDatagrams() Message ::  "[51] 7639500 18:05:21.7229 [Trace:HeartRateConnection] HeartRateConnection.HeartRateDataUpdate heart rate value changed. Value: 70 Device: id"

**Comment by cagnulein** [comment]
maybe a temporary issue on the ifit tablet? try again, you have to see that lines in the log to get it to the QZ main app

**Comment by samkenw** [comment]
Alright, will try again tomorrow.

**Comment by samkenw** [comment]
Couldn't test it because the treadmill didn't connect to the chest strap. 

However, I tried using a gpx file. Generally it's working! Before I simplified the file and de-noised it but the incline kept changing a lot. Is the app respecting the height profile in the gpx file or is it getting the height data from the map? Is there a setting for this? (My gpx has incline changes only every 500m)

**Comment by cagnulein** [comment]
> Couldn't test it because the treadmill didn't connect to the chest strap. 

haha ok :( 

> However, I tried using a gpx file. Generally it's working! Before I simplified the file and de-noised it but the incline kept changing a lot. Is the app respecting the height profile in the gpx file or is it getting the height data from the map? Is there a setting for this? (My gpx has incline changes only every 500m) 

Great! qz is getting the inclination directly from the gpx file. Let me know if you have any other questions!

**Comment by samkenw** [comment]
Will test the HR again later.

> Great! qz is getting the inclination directly from the gpx file. Let me know if you have any other questions!

Then there may be something wrong with the setting of the inclination? Because my file only has inclination changes every 500m and the incline was being adjusted constantly) 

This is what it should be:
![image](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/49457066/015485b8-6929-468f-8f3c-b8a6d6c9d7e2)

And this is what it looked like at the first minute:
![Mi  Nov  15 21_26_10 2023_speedInclination](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/49457066/9271939c-791c-4a50-aa69-db183ee3425e)

I can see in the logs that the app is trying to set the incline to the same value always (as expected) however it can't seem to achieve this in the treadmill so it is constantly trying to correct the incline. I suppose this is then an issue in that we can't set the incline exactly enough in the treadmill... (stupid sliders) do you have a suggestion to solve this?

**Comment by cagnulein** [comment]
please share the full log

**Comment by cagnulein** [comment]
got the issue @samkenw as you probably would like to say, i'm changing the inclination for every small changes, instead, i guess you treadmill is able to do only 0.5% changes, right? If so let me know and I will add a filter for this. Also what about speed? also 0.5 increments only?

**Comment by cagnulein** [comment]
[debug-Mi__Nov__15_20_11_47_2023.log.zip](https://github.com/cagnulein/QZCompanionNordictrackTreadmill/files/13374036/debug-Mi__Nov__15_20_11_47_2023.log.zip)
for my reference

**Comment by samkenw** [comment]
> got the issue @samkenw as you probably would like to say, i'm changing the inclination for every small changes, instead, i guess you treadmill is able to do only 0.5% changes, right? If so let me know and I will add a filter for this. Also what about speed? also 0.5 increments only?

Yes, I think this would solve the issue. (Inclination has 0,5 increments - speed has 0,1 increments) But I suppose ideally the user could set the filter values themselves? - this actually brings me to another great feature: Being able to set in the gpx settings "Only change incline if the incline difference is more than X compared to the current state". or something similar. 

Btw. I had the issue again, that at some point the companion stopped sending updates (fixed by just bringing the companion to the foreground for a second) - did you already address this issue? 

(Btw. Let me know when to open a new ticket for other issues - I will keep continuing to report issues/ ideas etc. but I don't expect these to be dealt with instantly - no pressure from me - I just find the app great and would like to help improve it)

**Comment by cagnulein** [comment]
> Yes, I think this would solve the issue. (Inclination has 0,5 increments - speed has 0,1 increments)

let's do this first, because in any case, the actual implementation doesn't have sense to send 0.1 increments for the inclination ;)



> But I suppose ideally the user could set the filter values themselves? - this actually brings me to another great feature: Being able to set in the gpx settings "Only change incline if the incline difference is more than X compared to the current state". or something similar.

of course I can add it to qz, but I'm asking if it's not already be possible outside qz, with some web app or similar? what do you think? @jonathancolledge @bepo7012 what do you think too?

> Btw. I had the issue again, that at some point the companion stopped sending updates (fixed by just bringing the companion to the foreground for a second) - did you already address this issue?

mmm not as I know. @victorypoint did you ever see this? actually the "alarm" code should bring the app alive again, isn't it?



> (Btw. Let me know when to open a new ticket for other issues - I will keep continuing to report issues/ ideas etc. but I don't expect these to be dealt with instantly - no pressure from me - I just find the app great and would like to help improve it)

you're doing great! i love to hear feedbacks from the users and if I can, following them in the development!

**Comment by cagnulein** [comment]
@samkenw modification done (filtering 0.5% inclination), I will send you ASAP

**Comment by Bepo7012** [comment]
Although i am no Fan of getting a "Option Forrest" i think making it setable would be a good idea. I know that Strava is already doing some kind of smoothing the inclination but i also see inclination changes quite often. Even after Strava's smothenings i sometimes see strange inclination jumps from 1.3 to 6 and back to 1 during a few seconds. Maybe a filter for such Peak Jumps (which can only be gpx Recording Errors or Errors from the Strava Algorithm) would be a good Idea too.

By now i did not find any other satisfying Web Services with the possibility to add or smoothen existing Inclination Data. DEM Databases may be helpful but the Resolution is mostly quite low within this Databases (30 - 90 Meters).

**Comment by victorypoint** [comment]
@samkenw, @cagnulein, I haven't come across this issue yet where I've had to bring companion to the foreground to start sending UDP packets. It's always worked fine in the background.

**Comment by samkenw** [comment]
> Although i am no Fan of getting a "Option Forrest" i think making it setable would be a good idea. I know that Strava is already doing some kind of smoothing the inclination but i also see inclination changes quite often. Even after Strava's smothenings i sometimes see strange inclination jumps from 1.3 to 6 and back to 1 during a few seconds. Maybe a filter for such Peak Jumps (which can only be gpx Recording Errors or Errors from the Strava Algorithm) would be a good Idea too.

I have generated a smooth gpx that only has very specific incline changes, however QZ seems to interpolate somehow (seeing this through the "target incline" in the app. This has values that do not exist in my gpx file) so Im not sure if this will solve all of the issue. In my opinion, the incline change should be instant, lets say from 3% to 10% on the QZ side - the treadmill needs some time anyway to revamp up the incline. 

I have also noticed, that the app is constantly sending adb swipe actions - even if the incline is set to the correct value. This is actually sometime quite a problem as the touch action interferes with other actions that the user performs. Suggestion: if the incline is set to the target incline, don't send any more swipe events until it needs to change again.

> @samkenw, @cagnulein, I haven't come across this issue yet where I've had to bring companion to the foreground to start sending UDP packets. It's always worked fine in the background.

I still get this even after a reset and fresh install... however, I have a totally different thought: 
@cagnulein would it be feasible to completely skip the companion app and just use something like this instead: https://github.com/Swind/pure-python-adb (or whatever language you prefer) - just saying, skip the companion and do everything via adb from the app?

**Comment by cagnulein** [comment]
@samkenw yes it's what i'm already doing for some nordictrack treadmills. now that we have adb also for ios i can extend this. no problem!

i will do this in the next days

i will read the gpx topic later

**Comment by cagnulein** [comment]
> I have generated a smooth gpx that only has very specific incline changes, however QZ seems to interpolate somehow (seeing this through the "target incline" in the app. This has values that do not exist in my gpx file) so Im not sure if this will solve all of the issue. In my opinion, the incline change should be instant, lets say from 3% to 10% on the QZ side - the treadmill needs some time anyway to ramp up the incline.

yes there is a 100m average https://github.com/cagnulein/qdomyos-zwift/blob/aad888cdabd1579d8c01f58d4d1ae2fdb82b1d19/src/trainprogram.cpp#L720

why? because each point in the gpx has a different inclination so you will have a inclination change every seconds and it will to heating your treadmill a lot! So i added an average of 100m for this reason. For sure I can remove or customize the 100m thing, but I already fought this and this is the best solution, believe me.



> I have also noticed, that the app is constantly sending adb swipe actions - even if the incline is set to the correct value. This is actually sometime quite a problem as the touch action interferes with other actions that the user performs. Suggestion: if the incline is set to the target incline, don't send any more swipe events until it needs to change again.

there is already a 

`double inc = qRound(requestInclination / 0.5) * 0.5;
                if(inc != currentInclination().value()) {`

so this shouldn't happen. do you have a log with this issue?



> I still get this even after a reset and fresh install... however, I have a totally different thought:
> @cagnulein would it be feasible to completely skip the companion app and just use something like this instead: https://github.com/Swind/pure-python-adb (or whatever language you prefer) - just saying, skip the companion and do everything via adb from the app?

I will try to add the iOS adb on a thread as I did for windows in the next days

**Comment by cagnulein** [comment]
@samkenw about the ADB thread: are you able to run the xcode simple project that I sent you and replace the ls command with a "logcat" command and check if you can see realtime output?

**Comment by samkenw** [comment]
Reg. the incline average, I understand, that's fine. 

I will test again in the next few days reg. the redundant swipe commands and send a log. 
Same goes for the Xcode project and the real-time adb.

**Comment by cagnulein** [comment]
sounds goods, thanks!

Il giorno lun 20 nov 2023 alle 23:43 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
ADB Logcat from Xcode works.

**Comment by cagnulein** [comment]
could I ask @samkenw how do you see the lines from adb logcat? i mean i'm interested in keeping the adb logcat in tail, so we should receive lines in realtime while you're doing a workout: so the event function is called every time with a new line?

**Comment by cagnulein** [comment]
I checked the code and in my opinion it returns all the text in a block, my idea is to run the logcat every 1-2 seconds.
What do you think @samkenw  ? I don't know how many milliseconds the treadmill tablet takes to answer back

**Comment by samkenw** [comment]
Hm.. if you check out this: [pure-python-adb](https://pypi.org/project/pure-python-adb/ ) there is an example `def dump_logcat(connection)` which I used for testing - `data` is each new line of data coming in in real time... From my testing (running this code on the Mac) it seemed fast enough - couldn't make out any delay. 

> could I ask @samkenw how do you see the lines from adb logcat? i mean i'm interested in keeping the adb logcat in tail, so we should receive lines in realtime while you're doing a workout: so the event function is called every time with a new line?

I can't make out if its coming as blocks or line-by-line - but it was continuous, I didn't have to read multiple times - just once until I stopped it. I do believe this will work as expected and we don't need to fetch in intervals.

**Comment by cagnulein** [comment]
i'm using this approach on windows, but i can't on ios or at least not with
the current ios library. i will check if i can do a mod in the library
directly, thanks

Il giorno mar 21 nov 2023 alle 21:07 Sam K ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
ok good news, I saw that in the iOS library there is already a logcat function, now I have to check how it behaves. I will try to setup a android device with adb to check on my own.

I will keep you updated, if it works on my end, it will be very easy to implement it. The only thing that could be a problem is to run the adb logcat at the same with with the input swipe thing.

I will check it also

**Comment by cagnulein** [comment]
@samkenw I can't unfortunately use any of my android devices to test this (i have only android with version greater than 10), also I can't also use an android emulator (i don't know why, but the adb project crashes)

So I need to "use" you to check my idea.
Running the adb project, calling the logcat, i would like to understand if the event didResponse is fired everytime it has a new line to show, or only once. If, as I hope, it will fire everytime has a new line, could you please also run another command in parallel in the same project so we can check that 2 adb commands are already handled?

Let me know if you need help on this

**Comment by samkenw** [comment]
Sure thing, I can give it a try. Should I update the Xcode project that I have to test this?

**Comment by cagnulein** [comment]
yes i mean, you have to modify it. if you want i can add some print or
other to help you

Il giorno ven 24 nov 2023 alle 12:07 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Alright, I will see how I cope (no experience in Xcode) and get back to you if I need any help.

**Comment by samkenw** [comment]
I think it would be faster if you send me the code snippet of the didResponse method because I have no idea :D

**Comment by cagnulein** [comment]
ok pull the project and you will have this commit
https://github.com/cagnulein/adb-ios/commit/807c30a58823322e7c02dc6673f9f13e59074deb

i did 2 things: i create 1 button for logcat and 1 for swipe

what you should do:
- change the ip in the file to connect to your tablet
- connect with this project
- press the logocat button and in the output window of xcode you should see
the output in realtime
- press also the swipe button and check that you can press it while you're
still receiving the logcat
- then put the whole output window of xcode in a file and attach it here so
i can double check it

Thanks!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 24 nov 2023 alle ore 13:31 cagnulein ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Alright, did it. So swipe works. BUT as soon as I press "logcat" I don't see any output AND after that, swipe does not work anymore.

**Comment by cagnulein** [comment]
interesting: and if you press the logcat first?

**Comment by samkenw** [comment]
yea, as soon as I press logcat it blocks the rest as it seems. But I am also not seeing any output from logcat.

**Comment by cagnulein** [comment]
ok pull it again and try again let me know

**Comment by samkenw** [comment]
Now swipe always works but still not seeing output from logcat.

**Comment by cagnulein** [comment]
<img width="1179" alt="Screenshot 2023-11-24 at 15 27 43" src="https://github.com/cagnulein/QZCompanionNordictrackTreadmill/assets/48216/15e43ea6-3b25-4d63-a195-3a58c6ff4602">
in the bottom window aren't you seeing a current date and time thing when you press logcat?

**Comment by cagnulein** [comment]
when you tried logcat last time, did you do something different? i mean im
calling the logcat command and printing it to the console. what did you do?

Il giorno ven 24 nov 2023 alle 15:32 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
I set it to the text field in the app. I can try that again.

**Comment by cagnulein** [comment]
you mean that you saw the adb logcat output in the text field?


Il giorno ven 24 nov 2023 alle 15:42 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
yes

**Comment by cagnulein** [comment]
ok pull it again, i restored it. if it works in this way i can start the patch in qz, thanks

**Comment by samkenw** [comment]
Yea, this shows the output in the text field. However it is not following. I have to press logcat to refresh and then it seems to get a bunch of data (again) + the new lines.

**Comment by cagnulein** [comment]
yes tailing it it will block the swipe, so i have everytime to get all the
output and parse it

Il giorno ven 24 nov 2023 alle 15:52 Sam K ***@***.***> ha
scritto:

**Comment by samkenw** [comment]
Is this also the case if getting the logcat is in a separate thread?

**Comment by cagnulein** [comment]
yes it's what I would like to do. that's why i asked to press both buttons.
if you want you can try to press them multiple times at the same time as
quick as you can
Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno ven 24 nov 2023 alle ore 16:00 Sam K ***@***.***>
ha scritto:

**Comment by cagnulein** [comment]
patch created with adb in a thread. I will try to build a version for you today https://github.com/cagnulein/qdomyos-zwift/pull/1829

**Comment by samkenw** [comment]
Tested. Not getting any data from the treadmill. Logs sent.

**Comment by cagnulein** [comment]
@samkenw thanks for the log (there is any particular reason why you're not adding them here instead of sending them by email?).
I checked and the issue it seems that it stucked after the connect.
That's strange because it's working on my simulator.
I will send you another one later today with some more debug log to understand this
Thanks!

**Comment by cagnulein** [comment]
hi @samkenw any news?

**Comment by samkenw** [comment]
sorry for the delay, have been busy. Will have to get back to you after the weekend.

**Comment by cagnulein** [comment]
Sure, no rush, have a great weekend!

Il giorno ven 1 dic 2023 alle 19:35 Sam K ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
@samkenw are you still interested in?
